ebos/eclgenericoutputblackoilmodule.cc

```diff
@@ -944,50 +944,42 @@ setRestart(const data::Solution& sol,
     assert(!saturation_[oilPhaseIdx].empty());
     saturation_[oilPhaseIdx][elemIdx] = so;
 
-    if (!fluidPressure_.empty() && sol.has("PRESSURE"))
-        fluidPressure_[elemIdx] = sol.data("PRESSURE")[globalDofIndex];
-    if (!temperature_.empty() && sol.has("TEMP"))
-        temperature_[elemIdx] = sol.data("TEMP")[globalDofIndex];
-    if (!rs_.empty() && sol.has("RS"))
-        rs_[elemIdx] = sol.data("RS")[globalDofIndex];
-    if (!rsw_.empty() && sol.has("RSW"))
-        rsw_[elemIdx] = sol.data("RSW")[globalDofIndex];
-    if (!rv_.empty() && sol.has("RV"))
-        rv_[elemIdx] = sol.data("RV")[globalDofIndex];
-    if (!rvw_.empty() && sol.has("RVW"))
-        rvw_[elemIdx] = sol.data("RVW")[globalDofIndex];
-    if (!cPolymer_.empty() && sol.has("POLYMER"))
-        cPolymer_[elemIdx] = sol.data("POLYMER")[globalDofIndex];
-    if (!cFoam_.empty() && sol.has("FOAM"))
-        cFoam_[elemIdx] = sol.data("FOAM")[globalDofIndex];
-    if (!cSalt_.empty() && sol.has("SALT"))
-        cSalt_[elemIdx] = sol.data("SALT")[globalDofIndex];
-    if (!pSalt_.empty() && sol.has("SALTP"))
-        pSalt_[elemIdx] = sol.data("SALTP")[globalDofIndex];
-    if (!permFact_.empty() && sol.has("PERMFACT"))
-        permFact_[elemIdx] = sol.data("PERMFACT")[globalDofIndex];
-    if (!soMax_.empty() && sol.has("SOMAX"))
-        soMax_[elemIdx] = sol.data("SOMAX")[globalDofIndex];
-    if (!pcSwMdcOw_.empty() && sol.has("PCSWM_OW"))
-        pcSwMdcOw_[elemIdx] = sol.data("PCSWM_OW")[globalDofIndex];
-    if (!krnSwMdcOw_.empty() && sol.has("KRNSW_OW"))
-        krnSwMdcOw_[elemIdx] = sol.data("KRNSW_OW")[globalDofIndex];
-    if (!pcSwMdcGo_.empty() && sol.has("PCSWM_GO"))
-        pcSwMdcGo_[elemIdx] = sol.data("PCSWM_GO")[globalDofIndex];
-    if (!krnSwMdcGo_.empty() && sol.has("KRNSW_GO"))
-        krnSwMdcGo_[elemIdx] = sol.data("KRNSW_GO")[globalDofIndex];
-    if (!ppcw_.empty() && sol.has("PPCW"))
-        ppcw_[elemIdx] = sol.data("PPCW")[globalDofIndex];
-    if (!cMicrobes_.empty() && sol.has("MICROBES"))
-        cMicrobes_[elemIdx] = sol.data("MICROBES")[globalDofIndex];
-    if (!cOxygen_.empty() && sol.has("OXYGEN"))
-        cOxygen_[elemIdx] = sol.data("OXYGEN")[globalDofIndex];
-    if (!cUrea_.empty() && sol.has("UREA"))
-        cUrea_[elemIdx] = sol.data("UREA")[globalDofIndex];
-    if (!cBiofilm_.empty() && sol.has("BIOFILM"))
-        cBiofilm_[elemIdx] = sol.data("BIOFILM")[globalDofIndex];
-    if (!cCalcite_.empty() && sol.has("CALCITE"))
-        cCalcite_[elemIdx] = sol.data("CALCITE")[globalDofIndex];
+    auto assign = [elemIdx, globalDofIndex, &sol](ScalarBuffer& data,
+                                                  const std::string& name)
+    {
+        if (!data.empty() && sol.has(name)) {
+            data[elemIdx] = sol.data(name)[globalDofIndex];
+        }
+    };
+
+    const std::vector<std::pair<std::string, ScalarBuffer*>> fields = {
```

Comment:
Using constexpr would be good, but on the other hand it would make adding another element a bit more cumbersome. I assume and hope that array with wrong number of elements in the initializer fails? Up to you.

Reply:
There is nothing compile time about pointers to members so no consexpr here. And i think it the slight overhead of a dynamic sized array is worth it for easier addition / removal of entries here.

Reply:
> Using `constexpr` would be good, but on the other hand it would make adding another element a bit more cumbersome

Class template argument deduction!
```C++
constexpr auto fields = std::array {
    std::pair { "BIOFILM", &cBioFilm_ },
    // ...
    std::pair { "UREA", &cUrea_ },
};
```
That way we won't have to count the number of array elements.

Reply:
Fair enough but i still cannot constexpr.

Reply:
> Fair enough but i still cannot constexpr.

Why not?  Both `array` and `pair` have `constexpr` constructors.

Reply:
Yes but the data (pointer to members) is not known at compile time.

Reply:
okay, if you don't believe me, do you believe the compiler?
```
/home/akva/kode/opm/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:956:20: error: use of ‘this’ in a constant expression
  956 |     constexpr auto fields = std::array{
      |                    ^~~~~~
```

Reply:
> okay, if you don't believe me, do you believe the compiler?

Normally yes, but `&cBioFilm_` is **very** different from `&EclGenericOutputBlackoilModule::cBioFilm_`.  The latter is a pointer to a data member that only depends on the class.  The former is a pointer to a member that also depends on the object.  The current version is fine, but I think it'd be interesting to see if
```C++
std::array {
    std::pair { "BIOFILM", &EclGenericOutputBlackoilModule::cBioFilm_ },
    // ...
    std::pair { "UREA", &EclGenericOutputBlackoilModule::cUrea_ },
}
```
could be used in a `constexpr` context.

Reply:
well this was new to me. it works as such, however

```
/home/akva/kode/opm/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:983:37: error: invalid use of unary ‘*’ on pointer to member
  983 |                   { assign(p.first, *p.second); });
      |                                     ^~~~~~~~~
```

Reply:
> however: `error: invalid use of unary ‘*’ on pointer to member`

Yeah, like [I said](https://github.com/OPM/opm-simulators/pull/4776#discussion_r1285713260): You need different syntax and the parameter type changes too.  In particular, the `assign` lambda would be something like
```C++
auto assign = [this, elemIdx, globalDofIndex, &sol](const std::string& name,
                                                    auto* dataMemberPtr)
{
    auto& data = this->*dataMemberPtr;
    if (!data.empty() && sol.has(name)) {
        data[elemIdx] = sol.data(name)[globalDofIndex];
    }
};
```
and you'd call it with `p.second` rather than `*p.second`.

That said, in this situation I don't think the benefit of `constexpr` is high enough to justify the cost of the increased syntactical complexity so I'll merge the current version as is once the build check is green.